[SPARK-54305][SQL][PYTHON] Add admission control support to Python DataSource streaming API

jiteshsoni opened the pull request:
[SPARK-50604][PYTHON][SS] Support admission control for Python Streaming Data Source

### What changes were proposed in this pull request?

This PR implements admission control for Python Streaming Data Sources, specifically adding support for the `maxRecordsPerBatch` option. This allows users to limit the number of records processed per micro-batch, which is critical for rate limiting and stability in streaming applications.

The changes include:

1.  **Python API Updates**:
    *   Updated `DataSourceStreamReader.latestOffset` to accept an optional `start` offset argument: `def latestOffset(self, start: Optional[dict] = None)`.
    *   Updated `DataSourceStreamReader.latestOffset` to support returning a tuple of `(capped_offset, true_latest_offset)` alongside the existing single offset return type.
    *   Modified `_SimpleStreamReaderWrapper` to maintain compatibility with the new signature (though admission control is not supported for simple readers).

2.  **Scala/Java Engine Updates**:
    *   Updated `PythonMicroBatchStream` to implement `SupportsAdmissionControl`.
    *   Added logic to validate and parse `maxRecordsPerBatch`.
    *   Explicitly blocked `maxFilesPerBatch` and `maxBytesPerBatch` with informative error messages as they are not supported for Python sources.

3.  **Internal Protocol**:
    *   Introduced a new RPC command `LATEST_OFFSET_WITH_REPORT_FUNC_ID` (890) to fetch both the capped offset and the true latest offset in a single call from the Python worker to the JVM.

4.  **Documentation & Examples**:
    *   Updated `python/docs/source/tutorial/sql/python_data_source.rst` to explain how to implement admission control in custom data sources.
    *   Added a new example `examples/src/main/python/sql/streaming/structured_blockchain_admission_control.py` demonstrating the feature.

### Why are the changes needed?
Currently, Python Streaming Data Sources read all available data in every batch. Without admission control, a sudden backlog of data can cause OOMs or unstable processing times. Enabling `maxRecordsPerBatch` brings Python sources closer to parity with native Spark sources regarding rate limiting and resource management.

### Does this PR introduce _any_ user-facing change?
Yes.

*   **API Change**: The `latestOffset` method in `DataSourceStreamReader` now accepts an optional `start` parameter.
    *   *Backward Compatibility*: Existing implementations that do not accept arguments will continue to work (the runner handles `TypeError` and falls back to the no-arg call).
*   **New Feature**: Users can now use `.option("maxRecordsPerBatch", ...)` with Python data sources to limit batch sizes.
*   **Error Handling**: Users attempting to use `maxFilesPerBatch` or `maxBytesPerBatch` with Python sources will now receive a clear `IllegalArgumentException` instead of the option being ignored or causing undefined behavior.

### How was this patch tested?
*   **Unit Tests**: Added new tests in `PythonStreamingDataSourceSuite.scala`:
    *   `test("admission control: maxRecordsPerBatch with DataSourceStreamReader")`: Verifies that batch sizes are respected.
    *   Tests verifying that `IllegalArgumentException` is thrown for unsupported options (`maxFilesPerBatch`, `maxBytesPerBatch`).
*   **Manual Verification**: Verified using the new `structured_blockchain_admission_control.py` example script, confirming that the micro-batch execution respects the configured block limits.

### Was this patch authored or co-authored using generative AI tooling?
No.

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Add admission control support to Python DataSource streaming API

This PR adds admission control support to Python streaming data sources by:

1. Enhanced Python API: Modified DataSourceStreamReader.latestOffset() to accept optional start and limit parameters
2. Scala Bridge: Updated PythonMicroBatchStream to validate admission control options and throw IllegalArgumentException for invalid values
3. Python-JVM Communication: Added function IDs and handlers in PythonStreamingSourceRunner for admission control parameters
4. Python Worker: Implemented latest_offset_with_report_func() with backward compatibility fallback
5. Documentation: Added comprehensive guide in python_data_source.rst
6. Tests: Added validation tests for invalid admission control parameters
7. Example: Created structured_blockchain_admission_control.py demonstrating the Full API

Key Benefits:
- Predictable performance with controlled batch sizes
- Rate limiting and backpressure support
- Feature parity with Scala DataSource capabilities
- Full backward compatibility (all parameters optional)

jiteshsoni committed:
[SPARK-54305] Fix PythonMicroBatchStream latestOffset backward compatibility

Change latestOffset() from throwing UnsupportedOperationException to
bridging to latestOffset(Offset, ReadLimit) for backward compatibility.

This fixes test failures in PythonStreamingDataSourceSuite where the
old latestOffset() signature is still called by Spark's streaming
execution engine.

jiteshsoni committed:
[SPARK-54305] Fix PythonMicroBatchStream backward compatibility (take 2)

- Fix null handling in latestOffset bridge method
- Handle null start offset in PythonStreamingSourceRunner by converting to empty string
- Update Python side to convert empty string back to None
- Fix error action name to be 'latestOffset' instead of 'latestOffsetWithReport'

All PythonStreamingDataSourceSuite tests now pass (8/8).

jiteshsoni committed:
[SPARK-54305] Fix flake8 E501 line-too-long errors

Add noqa: E501 comments to lines where Black and flake8 conflict.
Black considers these lines correctly formatted, but they exceed 79 chars.

jiteshsoni committed:
Fix(lint): Resolve multiple Python linting and formatting issues

This commit addresses several issues flagged by the CI pipeline:

- Fixes Flake8 E501 (line too long) errors in 'structured_blockchain_admission_control.py' by refactoring docstrings and long lines.
- Applies Black formatting to 'datasource_internal.py' and 'python_streaming_source_runner.py' to resolve pre-existing formatting inconsistencies.

jiteshsoni committed:
reformatted file  with 'dev/reformat-python' script

jiteshsoni committed:
MyPy Fixes

HeartSaVioR commented on python/docs/source/tutorial/sql/python_data_source.rst:
```diff
@@ -487,6 +487,101 @@ We can also use the same data source in streaming reader and writer
+            true_latest = self.get_latest_offset()
+            
+            # Apply admission control if configured
+            if limit and limit.get("type") == "maxRows":
```
Maybe good to hear some opinion on this.

@allisonwang-db 
We are porting some param class hierarchy to python data source.

ReadLimit
+- ReadMaxRows
+- ReadMaxFiles
+- ReadMaxBytes
+- ReadAllAvailable

The interface ReadLimit does not actually mean implementations share the method to read the value. That's mostly a marker and caller figures out the actual type and handles it accordingly.

Here, @jiteshsoni simply picks up dict with type names and corresponding param name per type. Would it be python friendly port, or we have better way of porting leveraging typing in Python?

Thanks in advance!

HeartSaVioR commented on python/docs/source/tutorial/sql/python_data_source.rst:
```diff
@@ -487,6 +487,101 @@ We can also use the same data source in streaming reader and writer
+  - ``{"type": "maxFiles", "maxFiles": N}`` for file-based limits  
+  - ``{"type": "maxBytes", "maxBytes": N}`` for byte-based limits
+  - ``{"type": "allAvailable"}`` for no limit
+- **SimpleDataSourceStreamReader**: Users of the simple API don't need to implement this; the framework handles admission control automatically
```
I'm not following this sentence. Do you mean users can still provide the above option and the implementation of SimpleDataSourceStreamReader can handle it without requiring the data source itself to do it? How would it be possible? e.g. The data source reads from Kafka and the option is max files.

HeartSaVioR commented on python/pyspark/sql/datasource.py:
```diff
@@ -714,20 +714,46 @@ def initialOffset(self) -> dict:
         )
 
-    def latestOffset(self) -> dict:
+    def latestOffset(
```
Maybe good to clarify the case first. Let's assume user decided not to implement the admission control in the data source implementation. What is the preferred way to do it? Throwing exception if the type of limit is not "allAvailable"?

Also based on the explanation of parameter, it doesn't seem like we are providing the same with Scala API.

```
  /**
   * Returns the most recent offset available given a read limit. The start offset can be used
   * to figure out how much new data should be read given the limit. Users should implement this
   * method instead of latestOffset for a MicroBatchStream or getOffset for Source.
   * <p>
   * When this method is called on a `Source`, the source can return `null` if there is no
   * data to process. In addition, for the very first micro-batch, the `startOffset` will be
   * null as well.
   * <p>
   * When this method is called on a MicroBatchStream, the `startOffset` will be `initialOffset`
   * for the very first micro-batch. The source can return `null` if there is no data to process.
   */
  Offset latestOffset(Offset startOffset, ReadLimit limit);
```

Python Data Source API is technically bound to DSv2 (MicroBatchStream), and we will need to provide some value to `start` all the time.

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala:
```diff
@@ -49,37 +73,120 @@ class PythonMicroBatchStream(
+   * offset for reporting purposes.
+   *
+   * @param startOffset
+   *   the starting offset, may be null
```
nit: let's put in the same line if it doesn't exceed 100 chars

HeartSaVioR replied:
nit: Btw, let's not add documentation unless there is noticeable difference from the method doc of base method.

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala:
```diff
@@ -49,37 +73,120 @@ class PythonMicroBatchStream(
+    cappedOffset
+  }
+
+  /**
```
nit: Let's not add documentation unless there is noticeable difference from the method doc of base method.

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala:
```diff
@@ -28,15 +29,38 @@ import org.apache.spark.storage.{PythonStreamBlockId, StorageLevel}
+ *   - `maxFilesPerBatch`: Maximum number of files per batch (Int, must be > 0)
+ *   - `maxBytesPerBatch`: Maximum bytes per batch (Long, must be > 0)
+ *
+ * @param ds
```
nit: keep them in single line unless it exceeds 100 chars

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala:
```diff
@@ -49,37 +73,120 @@ class PythonMicroBatchStream(
+      }
+    }
+
+    if (options.containsKey("maxRecordsPerBatch")) {
```
It'd be preferred to use constant fields for constant strings - let's put constants in companion object.

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/streaming/PythonStreamingSourceRunner.scala:
```diff
@@ -99,8 +103,8 @@ class PythonStreamingSourceRunner(
     pythonWorker = Some(worker)
     pythonWorkerFactory = Some(workerFactory)
 
-    val stream = new BufferedOutputStream(
```
nit: Same here.

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/streaming/PythonStreamingSourceRunner.scala:
```diff
@@ -188,14 +182,130 @@ class PythonStreamingSourceRunner(
+    if (len == SpecialLengths.PYTHON_EXCEPTION_THROWN) {
+      val msg = PythonWorkerUtils.readUTF(dataIn)
+      throw QueryExecutionErrors.pythonStreamingDataSourceRuntimeError(
+        action = "latestOffset",
```
nit: maybe good to follow the existing style

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/streaming/PythonStreamingSourceRunner.scala:
```diff
@@ -188,14 +182,130 @@ class PythonStreamingSourceRunner(
+    if (len == SpecialLengths.PYTHON_EXCEPTION_THROWN) {
+      val msg = PythonWorkerUtils.readUTF(dataIn)
+      throw QueryExecutionErrors.pythonStreamingDataSourceRuntimeError(
+        action = "reportLatestOffset",
```
nit: maybe good to follow the existing style

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/streaming/PythonStreamingSourceRunner.scala:
```diff
@@ -188,14 +182,130 @@ class PythonStreamingSourceRunner(
+  /**
+   * Serializes a ReadLimit to JSON format for Python. Uses json4s for safe JSON construction.
+   *
+   * @param limit
```
nit: one liner, please

HeartSaVioR commented on python/pyspark/sql/streaming/python_streaming_source_runner.py:
```diff
@@ -88,25 +94,86 @@ def partitions_func(
+            else:
+                capped_offset = true_latest_offset = fallback_result
+        except Exception as fallback_error:
+            raise IllegalArgumentException(
```
I'm not very sure this is a representative exception. Shall we not wrap the exception and throw it as it is? If you want to give the info of operation into exception, it's OK to wrap but IllegalArgumentException is too narrow one to cover "Exception".

HeartSaVioR reviewed (commented):
First pass, not gone through the test and example since I have a couple major concerns/suggestions on API.

I see a bunch of changes to be not related to the actual change. I hope we try to reduce the overall code diff to help other reviewers to go through fewer code changes. I suspect 20-30% of code changes are just style preferences; let's not change it.

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Address review nits for Python streaming DataSource admission control

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Fix mypy override for SimpleStreamReaderWrapper.latestOffset

jiteshsoni committed:
Revert formatting in streaming source runners

jiteshsoni committed:
Adjust formatting in python streaming source runner

jiteshsoni committed:
Tweak docs for PythonStreamingSourceRunner

jiteshsoni committed:
Format PythonStreamingSourceRunner header

jiteshsoni committed:
Reformat Arrow vector conversion

jiteshsoni commented on python/pyspark/sql/datasource_internal.py:
```diff
@@ -88,14 +88,65 @@ def initialOffset(self) -> dict:
+                # Cap the data to max_rows
+                capped_data = data_list[:max_rows]
+                capped_iter = iter(capped_data)
+                # Calculate capped offset based on how many rows we're actually taking
```
Accepted your suggestion and removing admission control from _SimpleStreamReaderWrapper

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Follow existing style in PythonStreamingSourceRunner

jiteshsoni committed:
Revert "[SPARK-54305][SQL][PYTHON] Follow existing style in PythonStreamingSourceRunner"

This reverts commit a9a076dc69d4da7b3719e3e66568ee7a75bd24d2.

jiteshsoni committed:
Cleanup formatting-only changes

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Fix latestOffset override in SimpleStreamReaderWrapper

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Delegate MicroBatchStream.latestOffset

jiteshsoni commented on python/pyspark/sql/datasource.py:
```diff
@@ -714,20 +714,46 @@ def initialOffset(self) -> dict:
+        Parameters (optional - added in Spark 4.2 for admission control)
+        -------------------------------------------------------------------
+        start : dict, optional
+            The starting offset. Enables admission control when provided.
```
Removed the limit parameter

jiteshsoni committed:
[SPARK-54305][SQL][PYTHON] Minimize diff in PythonStreamingDataSourceSuite

jiteshsoni committed:
Follow existing style in latestOffsetWithReport

jiteshsoni committed:
Revert formatting churn in PythonStreamingDataSourceSuite

jiteshsoni committed:
Revert formatting in PythonStreamingDataSourceSuite

jiteshsoni committed:
Revert formatting churn in PythonStreamingDataSourceSuite

jiteshsoni committed:
Revert remaining formatting in PythonStreamingDataSourceSuite

jiteshsoni committed:
Update PythonStreamingDataSourceSuite formatting

jiteshsoni committed:
Fix minor formatting in PythonStreamingDataSourceSuite

jiteshsoni committed:
Fix @since for latestOffsetWithReport

jiteshsoni committed:
Fail fast on unsupported admission control options

jiteshsoni committed:
Use constants for admission control option keys

jiteshsoni committed:
Remove unreachable ReadLimit branches

jiteshsoni committed:
Avoid null ambiguity for startOffset json

jiteshsoni committed:
Reorder PythonMicroBatchStream import

jiteshsoni committed:
Fix formatting and constants imports

jiteshsoni committed:
1 line fix

jiteshsoni committed:
Fix mypy errors in datasource_internal.py

- Update latestOffset() signature to accept optional 'start' parameter to match parent class
- Change read() parameter name from 'input_partition' to 'partition' to match parent class
- Add runtime type assertion in read() for SimpleInputPartition
- Apply black formatting

jiteshsoni committed:
Apply black formatting via dev/reformat-python

- Compact function signature and list comprehension formatting
- Ensure CI lint-python checks pass

jiteshsoni committed:
Fix admission control validation to fail fast

- Move unsupported options validation (maxFilesPerBatch, maxBytesPerBatch) to stream initialization
- This ensures IllegalArgumentException is thrown during .load() not later during execution
- Fixes failing tests in PythonStreamingDataSourceSuite

jiteshsoni committed:
Solve the 2 argument handling test cases
